risc0/zkvm/src/host/protos/api.proto

```diff
@@ -0,0 +1,240 @@
+syntax = "proto3";
+
+import "google/protobuf/empty.proto";
+
+package protos.api;
+
+message ServerRequest {
+  oneof kind {
+    ExecuteRequest execute = 1;
+    ProveRequest prove = 2;
+    ProveSegmentRequest prove_segment = 3;
+    LiftRequest lift = 4;
+    JoinRequest join = 5;
+  }
+}
+
+message HelloRequest {
+  uint32 version = 1;
+}
+
+message HelloReply {
+  oneof kind {
+    HelloResult ok = 1;
+    GenericError error = 2;
+  }
+}
+
+message HelloResult {
+  uint32 version = 2;
+}
+
+message ExecuteRequest {
+  ExecutorEnv env = 1;
+  AssetRequest segments_out = 2;
+}
+
+message ProveRequest {
+  ExecutorEnv env = 1;
+  ProverOpts opts = 2;
+  AssetRequest receipt_out = 3;
+}
+
+message ProveSegmentRequest {
+  ProverOpts opts = 1;
+  Asset segment = 2;
+  AssetRequest receipt_out = 3;
+}
+
+message ProveSegmentReply {
+  oneof kind {
+    ProveSegmentResult ok = 1;
+    GenericError error = 2;
+  }
+}
+
+message LiftRequest {
+  ProverOpts opts = 1;
+  Asset receipt = 2;
+  AssetRequest receipt_out = 3;
+}
+
+message LiftReply {
+  oneof kind {
+    LiftResult ok = 1;
+    GenericError error = 2;
+  }
+}
+
+message LiftResult {
+  Asset receipt = 1;
+}
+
+message JoinRequest {
+  ProverOpts opts = 1;
+  Asset left = 2;
+  Asset right = 3;
+  AssetRequest receipt_out = 4;
+}
+
+message JoinReply {
+  oneof kind {
+    JoinResult ok = 1;
+    GenericError error = 2;
+  }
+}
+
+message JoinResult {
+  Asset receipt = 1;
+}
+
+message ExecutorEnv {
+  Binary binary = 1;
+  map<string, string> env_vars = 2;
+  repeated string slice_ios = 3;
+  repeated uint32 read_fds = 4;
+  repeated uint32 write_fds = 5;
+}
+
+message Binary {
+  enum Kind {
+    UNSPECIFIED = 0;
+    IMAGE = 1;
+    ELF = 2;
+  }
+
+  Kind kind = 1;
+  Asset asset = 2;
+}
+
+message ProverOpts {
+  string hashfn = 1;
+  bool lift = 2;
+}
+
+message SessionInfo {
+  uint32 segments = 1;
+  bytes journal = 2;
+  ExitCode exit_code = 3;
+}
+
+message SegmentInfo {
+  uint32 index = 1;
+  uint32 po2 = 2;
+  uint32 insn_cycles = 3;
+  Asset segment = 4;
+}
+
+message ProveSegmentResult {
+  Asset receipt = 1;
+}
+
+message ExitCode {
+  oneof kind {
+    uint32 halted = 1;
+    uint32 paused = 2;
+    google.protobuf.Empty system_split = 3;
+    google.protobuf.Empty session_limit = 4;
+  }
+}
+
+message Asset {
+  oneof kind {
+    bytes inline = 1;
+    string path = 2;
+  }
+}
+
+message AssetRequest {
+  oneof kind {
+    google.protobuf.Empty inline = 1;
+    string path = 2;
+  }
+}
+
+message ClientCallback {
```

Comment:
From reading this my assumption is that this is the message sent from the server to client each time a callback triggering event occurs. Is that accurate? If so, when would the `oneof` be an error?

Reply:
From reading further my assumption is that error would be returned as a final message in the stream if an error occurs. Maybe it would make sense to factor this as a final message that is either "ok" or "error"?

Reply:
The issue is, I need to know a priori which type to decode from the stream. I suppose we could do:
```
message ClientResult {
  oneof kind {
    ClientCallback ok = 1;
    GenericError error = 2;
  }
}
```

Reply:
To me that has slightly more clear semantics, but looking at this again it seems simple enough that the steam will contain zero or more `ClientCallback` messages and that one of them may be an error if something went wrong.